sse/sse.go

```diff
@@ -0,0 +1,156 @@
+package sse
+
+import (
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/missionMeteora/apiserv"
+)
+
+var (
+	RespNotAFlusher = apiserv.NewJSONErrorResponse(http.StatusInternalServerError, ErrNotAFlusher)
+)
+
+type dataChan chan []byte
+
+type multiStream struct {
+	clients map[dataChan]struct{}
+	mux     sync.Mutex
+	data    chan []byte
+}
+
+func (ms *multiStream) add(ch dataChan) {
+	ms.mux.Lock()
+	ms.clients[ch] = struct{}{}
+	ms.mux.Unlock()
+}
+
+func (ms *multiStream) remove(ch dataChan) (isEmpty bool) {
+	ms.mux.Lock()
+	delete(ms.clients, ch)
+	isEmpty = len(ms.clients) == 0
+	ms.mux.Unlock()
+	close(ch)
+
+	return
+}
+
+func (ms *multiStream) close() {
+	close(ms.data)
+}
+
+func (ms *multiStream) process() {
+	for b := range ms.data {
+		if b == nil {
+			return
+		}
+
+		for ch := range ms.clients {
+			if !trySend(ch, b) {
+				delete(ms.clients, ch)
+			}
+		}
+	}
+}
+
+func NewRouter() *Router {
+	return &Router{
+		mss: make(map[string]*multiStream, 8),
+	}
+}
+
+type Router struct {
+	mss map[string]*multiStream
+	mux sync.RWMutex
+}
+
+func (r *Router) getOrMake(id string) (ms *multiStream) {
+	r.mux.Lock()
+	if ms = r.mss[id]; ms == nil {
+		ms = &multiStream{
+			clients: make(map[dataChan]struct{}, 8),
+			data:    make(chan []byte),
+		}
+		go ms.process()
```

Comment:
Hmm.. if we call a goroutine within a lock..this getOrMake func ends..would the Lock still hold?